intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/KotlinMethods.kt

```diff
@@ -0,0 +1,137 @@
+    override fun removeEldestEntry(eldest: MutableMap.MutableEntry<MethodLocation, Boolean>?): Boolean {
+      return size > CAPACITY
+    }
+  }
+
+  /**
+   * Identify a Kotlin default method.
+   *
+   * A kotlin default method is a method that has a default implementation, but kotlin compiles
+   * in the inheritor an override that calls the default implementation (inner) class known as
+   * `DefaultImpls`.
+   *
+   * Ignoring labels, line numbers, and kotlin intrinsics (to check nullness), there's only a
+   * handful of interesting bytecode that invoke the static method in `{TheInterfaceType}$DefaultImpls`
+   * loading first `this`, then loading parameters if any.
+   *
+   * E.g. with
+   *
+   * ```
+   * // access flags 0x1
+   * public internalArgsReturningInternal(Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
+   *    L0
+   *     ALOAD 1
+   *     LDC "anInternalType"
+   *     INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullParameter (Ljava/lang/Object;Ljava/lang/String;)V
+   *     ALOAD 2
+   *     LDC "s"
+   *     INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullParameter (Ljava/lang/Object;Ljava/lang/String;)V
+   *    L1
+   *     LINENUMBER 5 L1
+   *     ALOAD 0
+   *     ALOAD 1
+   *     ALOAD 2
+   *     INVOKESTATIC internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.internalArgsReturningInternal (Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
+   *     ARETURN
+   *    L2
+   *     LOCALVARIABLE this Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI; L0 L2 0
+   *     LOCALVARIABLE anInternalType Linternal/defaultMethod/AnInternalType; L0 L2 1
+   *     LOCALVARIABLE s Ljava/lang/String; L0 L2 2
+   *     MAXSTACK = 3
+   *     MAXLOCALS = 3
+   * ```
+   *
+   * @see [`JvmDefault` annotation](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/src/kotlin/jvm/JvmDefault.kt)
+   */
+  fun Method.isKotlinDefaultMethod(): Boolean {
+    val method: Method = this
+    return cache.computeIfAbsent(method.location) {
+      return@computeIfAbsent isKotlinMethodInvokingDefaultImpls(method)
+    }
+  }
+
+  private fun Method.isKotlinMethodInvokingDefaultImpls(method: Method): Boolean {
+    // filter non kotlin classes
+    if (!method.containingClassFile.annotations.any { it.desc == "Lkotlin/Metadata;" }) {
+      return false
+    }
+
+    // Sanity check : if the method does not have bytecode
+    // this heuristic cannot run
+    if (instructions.isEmpty()) {
+      return false
+    }
+
+    // skip non opcodes, and kotlin intrinsics on arguments
+    val candidateOpcodes = mutableListOf<AbstractInsnNode>()
+    var i = 0
+    do {
+      val currentInsnNode = instructions[i]
+      if (currentInsnNode.opcode == -1) {
```

Comment:
I don't know whether this instruction list contains debug info, but if it does, maybe the whole loop can be simplified to skipping until the first LINENUMBER instruction?

Reply:
I don't think all plugins have debug info, I know that some are using obfuscators like pro-guard that may remove these (but of course keep parent symbols intact).
I think it's safer to assume debug info is not always present.